File: mobile/ios/Air/SubModules/UIHome/ViewControllers/Home/HomeVM.swift
//
//  HomeVM.swift
//  WalletContext
//
//  Created by Sina on 3/20/24.
//

import Foundation
import UIKit
import WalletContext
import WalletCore
import WReachability
import UIEarn

private let log = Log("HomeVM")

private let UPDATING_DELAY = 2


protocol HomeVMDelegate: AnyObject {
    func update(state: UpdateStatusView.State, animated: Bool)
    func updateBalance(balance: Double?, balance24h: Double?, walletTokens: [MTokenBalance])
    func changeAccountTo(accountId: String, isNew: Bool) async
    func transactionsUpdated(accountChanged: Bool, isUpdateEvent: Bool)
    func tokensChanged()
    func forceReload()
    func scrollToTop()
}

class HomeVM {
    
    weak var homeVMDelegate: HomeVMDelegate?
    
    let reachability = try! Reachability()
    var waitingForNetwork: Bool? = nil
    private var loadSwapAssetsTask: Task<Void, Never>?
    
    private var prevUpdatingState: UpdateStatusView.State? = nil
    private var setUpdatingAfterDelayTask: Task<Void, Never>? = nil
    
    deinit {
        reachability.stopNotifier()
    }
    
    init(homeVMDelegate: HomeVMDelegate) {
        self.homeVMDelegate = homeVMDelegate
        WalletCoreData.add(eventObserver: self)

        // Listen for network connection events
        reachability.whenReachable = { [weak self] reachability in
            guard let self else {return}
            if waitingForNetwork == true {
                waitingForNetwork = false
                refreshTransactions()
            } else {
                waitingForNetwork = false
                updateStatus()
            }
        }
        reachability.whenUnreachable = { [weak self] _ in
            self?.waitingForNetwork = true
            self?.updateStatus()
        }
        do {
            try reachability.startNotifier()
        } catch {
        }
    }

    private func updateStatus() {
        guard waitingForNetwork == false else {
            homeVMDelegate?.update(state: .waitingForNetwork, animated: true)
            prevUpdatingState = .waitingForNetwork
            return
        }
        if AccountStore.updatingActivities || AccountStore.updatingBalance {
            if prevUpdatingState == .waitingForNetwork || prevUpdatingState == nil {
                log.info("updateStatus - network connected - updating", fileOnly: true)
                self.prevUpdatingState = .updating
                self.homeVMDelegate?.update(state: .updating, animated: true)
                setUpdatingAfterDelayTask?.cancel()
                return
            }
            if setUpdatingAfterDelayTask == nil || setUpdatingAfterDelayTask?.isCancelled == true {
                self.setUpdatingAfterDelayTask = Task { [self] in
                    do {
                        try await Task.sleep(for: .seconds(UPDATING_DELAY))
                        try Task.checkCancellation()
                        if AccountStore.updatingActivities || AccountStore.updatingBalance {
                            log.info("setUpdatingAfterDelayTask executing - \(UPDATING_DELAY)s passed", fileOnly: true)
                            self.prevUpdatingState = .updating
                            self.homeVMDelegate?.update(state: .updating, animated: true)
                        } else {
                            self.prevUpdatingState = .updated
                            self.homeVMDelegate?.update(state: .updated, animated: true)
                        }
                    } catch {
                        // canceled. no longer updating
                        log.info("setUpdatingAfterDelayTask is canceled", fileOnly: true)
                    }
                }
            }
        } else {
            log.info("updateStatus - updated", fileOnly: true)
            setUpdatingAfterDelayTask?.cancel()
            homeVMDelegate?.update(state: .updated, animated: true)
            prevUpdatingState = .updated
        }
    }
    
    // MARK: - Wallet Public Variables
    var account: MAccount? = nil
    
    // while balances are not loaded, do not show anything!
    var balancesLoaded: Bool {
        BalanceStore.currentAccountBalances.count > 0
    }
    
    var isGeneralDataAvailable: Bool {
        TokenStore.swapAssets != nil &&
        TokenStore.tokens.count > 1 &&
        balancesLoaded &&
        (BalanceStore.currentAccountBalances[TONCOIN_SLUG] != nil || BalanceStore.currentAccountBalances[TRX_SLUG] != nil)
    }
    
    // MARK: - Init wallet info
    func initWalletInfo() {
        // fetch all data
        let accountId = AccountStore.accountId!
        if let account = AccountStore.accountsById[accountId] {
            self.account = account
            // unique identifier to detect and ignore revoked requests
            /*if balancesLoaded != true {
                loadWalletBalances()
            }*/
            NftStore.forceLoad(for: account.id)
        } else {
            DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
                guard let self else {return}
                initWalletInfo()
            }
        }
        DispatchQueue.global(qos: .background).asyncAfter(deadline: .now() + 2) {
            EarnVM.sharedTon.preload()
//            EarnVM.sharedMycoin.preload()
        }
    }
    
    // called on pull to refresh / selected slug change / after network reconnection / when retrying failed tries
    func refreshTransactions(slugChanged: Bool = false) {
        // init requests
        initWalletInfo()
        // update token prices
        Api.tryUpdateTokenPrices()
    }

    func dataUpdated(transactions: Bool = true) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self, let accountId = AccountStore.accountId else { return }
            // make sure balances are loaded
            if !balancesLoaded {
                log.info("Balances not loaded yet")
                return
            }
            // make sure default event for receiving toncoin is also called
            if BalanceStore.balancesEventCalledOnce[accountId] == nil {
                log.info("balancesEventCalledOnce not loaded yet")
                return
            }
            // make sure assets are loaded
            if TokenStore.swapAssets == nil {
                log.info("swap assets are not loaded yet")
                DispatchQueue.main.asyncAfter(deadline: .now() + 5, execute: { [weak self] in
                    self?.loadSwapAssetsIfNeeded()
                })
            }
            // update balance view
            let accountBalanceData = BalanceStore.currentAccountBalanceData
            let walletTokens = Array(accountBalanceData?.walletTokens ?? [])
            DispatchQueue.main.async { [weak self] in
                guard let self else {return}
                // reload balance
                homeVMDelegate?.updateBalance(balance: accountBalanceData?.totalBalance,
                                              balance24h: accountBalanceData?.totalBalanceYesterday,
                                              walletTokens: walletTokens)
                // reload transaction prices
                homeVMDelegate?.tokensChanged()
            }
        }
    }
    
    private func loadSwapAssetsIfNeeded() {
        if TokenStore.swapAssets == nil && self.loadSwapAssetsTask == nil {
            self.loadSwapAssetsTask = Task { [weak self] in
                do {
                    _ = try await TokenStore.updateSwapAssets()
                } catch {
                    try? await Task.sleep(for: .seconds(5))
                    if !Task.isCancelled {
                        self?.loadSwapAssetsTask = nil
                        self?.loadSwapAssetsIfNeeded()
                    }
                }
            }
        }
    }
    
    func baseCurrencyChanged() {
        // reload tableview to make it clear as the tokens are not up to date
        homeVMDelegate?.tokensChanged()
        // make header empty like initialization view
        homeVMDelegate?.updateBalance(balance: nil,
                                      balance24h: nil,
                                      walletTokens: [])
    }

    fileprivate func accountChanged(isNew: Bool) {
        // reset load states, active network requests will also be ignored automatically
        guard let account = AccountStore.account else { return }
        self.account = AccountStore.account
        self.setUpdatingAfterDelayTask?.cancel()
        self.setUpdatingAfterDelayTask = nil
        homeVMDelegate?.update(state: waitingForNetwork == true ? .waitingForNetwork : .updated, animated: false)
        
        let accountBalanceData = BalanceStore.currentAccountBalanceData
        if let walletTokens = BalanceStore.currentAccountBalanceData?.walletTokens, walletTokens.count > 0 {
            DispatchQueue.main.async { [weak self] in
                guard let self else {return}
                // reload transaction prices
                homeVMDelegate?.transactionsUpdated(accountChanged: true, isUpdateEvent: false)
                // reload balance
                homeVMDelegate?.updateBalance(balance: accountBalanceData?.totalBalance,
                                              balance24h: accountBalanceData?.totalBalanceYesterday,
                                              walletTokens: walletTokens)
            }
        } else {
            homeVMDelegate?.updateBalance(balance: nil, balance24h: nil, walletTokens: [])
        }
        Task {
            await homeVMDelegate?.changeAccountTo(accountId: account.id, isNew: isNew)
        }
        // get all data again
        initWalletInfo()
        homeVMDelegate?.scrollToTop()
    }

}

extension HomeVM: WalletCoreData.EventsObserver {
    func walletCore(event: WalletCoreData.Event) {
        switch event {
        case .balanceChanged(let isFirstUpdate):
            dataUpdated()
            if isFirstUpdate {
                homeVMDelegate?.transactionsUpdated(accountChanged: false, isUpdateEvent: false)
            }
            break
        case .tokensChanged, .swapTokensChanged:
            dataUpdated()
            homeVMDelegate?.forceReload()
            break
        case .baseCurrencyChanged:
            baseCurrencyChanged()
        case .accountChanged(_, let isNew):
            accountChanged(isNew: isNew)
            break
        case .accountNameChanged:
            dataUpdated()
            break
        case .forceReload:
            homeVMDelegate?.forceReload()
            break
        case .assetsAndActivityDataUpdated:
            dataUpdated()
            homeVMDelegate?.forceReload()
        case .updatingStatusChanged:
            updateStatus()
        default:
            break
        }
    }
}

File: mobile/android/air/SubModules/UIHome/src/main/java/org/mytonwallet/uihome/home/HomeVM.kt
package org.mytonwallet.uihome.home

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import org.mytonwallet.app_air.walletcore.TONCOIN_SLUG
import org.mytonwallet.app_air.walletcore.TRON_SLUG
import org.mytonwallet.app_air.walletcore.WalletCore
import org.mytonwallet.app_air.walletcore.api.fetchAccount
import org.mytonwallet.app_air.walletcore.api.requestDAppList
import org.mytonwallet.app_air.walletcore.api.swapGetAssets
import org.mytonwallet.app_air.walletcore.api.tryUpdatePrices
import org.mytonwallet.app_air.walletcore.helpers.ActivityLoader
import org.mytonwallet.app_air.walletcore.helpers.IActivityLoader
import org.mytonwallet.app_air.walletcore.stores.AccountStore
import org.mytonwallet.app_air.walletcore.stores.BalanceStore
import org.mytonwallet.app_air.walletcore.stores.TokenStore
import org.mytonwallet.uihome.home.views.UpdateStatusView
import java.lang.ref.WeakReference

class HomeVM(val context: Context, delegate: Delegate) : WalletCore.EventObserver,
    IActivityLoader.Delegate {

    interface Delegate {
        fun update(state: UpdateStatusView.State, animated: Boolean)
        fun updateBalance(balance: Double?, balance24h: Double?, accountChanged: Boolean)
        fun reloadCard()
        fun updateEmptyView()

        // instant update list
        fun transactionsLoaded()

        // animated update transactions
        fun transactionsUpdated(isUpdateEvent: Boolean)
        fun cacheNotFound()
        fun loadedAll()

        fun loadStakingData()
        fun stakingDataUpdated()

        // fun forceReload()
        fun instantScrollToTop()

        fun updateActionsView()
        fun reloadTabs(accountChanged: Boolean)
        fun accountNameChanged()
        fun accountConfigChanged()
    }

    val delegate: WeakReference<Delegate> = WeakReference(delegate)

    var waitingForNetwork = false

    // unique identifier to detect and ignore revoked requests
    internal var accountCode = 0L

    // Activities variables
    internal var activityLoaderHelper: IActivityLoader? = null
    internal var calledReady = false

    // loaded data
    private val balancesLoaded: Boolean
        get() {
            return !BalanceStore.getBalances(accountId = AccountStore.activeAccountId)
                .isNullOrEmpty()
        }

    private val handler = Handler(Looper.getMainLooper())
    private val updateRunnable = object : Runnable {
        override fun run() {
            checkUpdatingTimer = null
            updateStatus()
        }
    }
    private var checkUpdatingTimer: Runnable? = null
    private fun startUpdatingTimer() {
        if (checkUpdatingTimer == null) {
            checkUpdatingTimer = updateRunnable
            handler.postDelayed(updateRunnable, 2000)
        }
    }

    private fun stopUpdatingTimer() {
        checkUpdatingTimer?.let { handler.removeCallbacks(it) }
        checkUpdatingTimer = null
    }

    fun delegateIsReady() {
        WalletCore.registerObserver(this)
        if (!WalletCore.isConnected()) {
            connectionLost()
        }
        startUpdatingTimer()
    }

    val isGeneralDataAvailable: Boolean
        get() {
            return TokenStore.swapAssets != null &&
                TokenStore.loadedAllTokens &&
                BalanceStore.getBalances(AccountStore.activeAccountId)
                    ?.get(TONCOIN_SLUG) != null
        }

    // Called on start or account change
    fun initWalletInfo() {
        // fetch all data
        val accountId = AccountStore.activeAccountId ?: return
        WalletCore.fetchAccount(accountId) { account, err ->
            accountCode = System.currentTimeMillis()
            activityLoaderHelper?.clean()
            activityLoaderHelper =
                ActivityLoader(context, accountId, null, WeakReference(this))
            activityLoaderHelper?.askForActivities()
        }
        // Load staking data
        delegate.get()?.loadStakingData()

        WalletCore.requestDAppList()
    }

    // called on pull to refresh / selected slug change / after network reconnection / when retrying failed tries
    private fun HomeVM.refreshTransactions() {
        // init requests
        initWalletInfo()
        // update token prices
        WalletCore.tryUpdatePrices()
    }

    private fun dataUpdated() {
        // make sure balances are loaded
        if (!balancesLoaded) {
            Log.i("HomeVM", "Balances not loaded yet")
            return
        }

        // make sure tokens are loaded
        if (!TokenStore.loadedAllTokens) {
            Log.i("HomeVM", "tokens not loaded yet")
            /*
             to prevent over-requesting for prices, we wait for the initial request and response of the js logic;
             to track tokens load, wait for for 5 seconds, if not loaded yet, it might be an issue initializing it in js, let's manually reload it.
             we only call `tryUpdateTokenPrices` on manual refresh, network reconnect and base currency change!
             */
            Handler(Looper.getMainLooper()).postDelayed({
                if (!TokenStore.loadedAllTokens) {
                    WalletCore.tryUpdatePrices()
                }
            }, 5000)
            return
        }

        // make sure default event for receiving toncoin is also called
        val balances = BalanceStore.getBalances(AccountStore.activeAccountId)
        if (
            balances?.get(TONCOIN_SLUG) == null ||
            (AccountStore.activeAccount?.isMultichain == true && balances.get(TRON_SLUG) == null)
        ) {
            Log.i("HomeVM", "balancesEventCalledOnce not loaded yet")
            return
        }

        // make sure assets are loaded
        if (TokenStore.swapAssets == null) {
            Log.i("HomeVM", "swap assets are not loaded yet")
            Handler(Looper.getMainLooper()).postDelayed({
                if (TokenStore.swapAssets == null) {
                    WalletCore.swapGetAssets(true) { assets, err ->
                        dataUpdated()
                    }
                }
            }, 5000)
            return
        }

        updateBalanceView(false)

        delegate.get()?.transactionsLoaded()
        delegate.get()?.updateEmptyView()
    }

    private fun updateBalanceView(accountChanged: Boolean) {
        if (!balancesLoaded || TokenStore.getToken(TONCOIN_SLUG)?.price == null) {
            delegate.get()?.updateBalance(null, null, accountChanged)
            return
        }

        // update balance view
        val walletTokens = AccountStore.assetsAndActivityData.getAllTokens().toMutableList()

        val totalBalance = walletTokens.sumOf { it.toBaseCurrency ?: 0.0 }
        val totalBalanceYesterday = walletTokens.sumOf { it.toBaseCurrency24h ?: 0.0 }

        // reload balance
        delegate.get()?.updateBalance(
            balance = totalBalance,
            balance24h = totalBalanceYesterday,
            accountChanged
        )
    }

    private fun baseCurrencyChanged() {
        delegate.get()?.updateBalance(null, null, false)
        // show loading again
        delegate.get()?.updateEmptyView()
        // reload tableview to make it clear as the tokens are not up to date
        delegate.get()?.transactionsLoaded()
        // make header empty like initialization view
        updateBalanceView(false)
    }

    private fun updateStatus(animated: Boolean = true) {
        if (waitingForNetwork) {
            // It's either `waiting for network` or `not specified` yet!
            return
        }
        if (AccountStore.updatingActivities || AccountStore.updatingBalance) {
            delegate.get()?.update(UpdateStatusView.State.Updating, animated)
        } else {
            delegate.get()?.update(UpdateStatusView.State.Updated, animated)
        }
    }

    private fun connectionLost() {
        waitingForNetwork = true
        delegate.get()?.update(UpdateStatusView.State.WaitingForNetwork, true)
    }

    private fun accountChanged() {
        calledReady = false

        activityLoaderHelper?.clean()
        activityLoaderHelper = null
        // reload tableview to make it clear as the tokens are not up to date
        delegate.get()?.transactionsLoaded()
        // get all data again
        initWalletInfo()
        // make header empty like initialization view
        updateBalanceView(true)
        delegate.get()?.instantScrollToTop()
        updateStatus(false)

        // update actions view
        delegate.get()?.updateActionsView()
        delegate.get()?.reloadTabs(true)
        delegate.get()?.accountNameChanged()
        delegate.get()?.accountConfigChanged()

        dataUpdated()
    }

    override fun onWalletEvent(event: WalletCore.Event) {
        when (event) {
            WalletCore.Event.BalanceChanged, WalletCore.Event.TokensChanged -> {
                dataUpdated()
            }

            WalletCore.Event.BaseCurrencyChanged -> {
                baseCurrencyChanged()
            }

            is WalletCore.Event.AccountChanged -> {
                accountChanged()
            }

            WalletCore.Event.AccountNameChanged -> {
                delegate.get()?.accountNameChanged()
                dataUpdated()
            }

            WalletCore.Event.StakingDataUpdated -> {
                delegate.get()?.stakingDataUpdated()
            }

            WalletCore.Event.AssetsAndActivityDataUpdated -> {
                dataUpdated()
            }

            WalletCore.Event.NetworkConnected -> {
                if (waitingForNetwork) {
                    waitingForNetwork = false
                    refreshTransactions()
                } else {
                    waitingForNetwork = false
                    updateStatus()
                }
            }

            WalletCore.Event.NetworkDisconnected -> {
                connectionLost()
            }

            WalletCore.Event.NftCardUpdated -> {
                delegate.get()?.reloadCard()
            }

            WalletCore.Event.NftsUpdated, WalletCore.Event.HomeNftCollectionsUpdated -> {
                delegate.get()?.reloadTabs(false)
            }

            WalletCore.Event.UpdatingStatusChanged -> {
                startUpdatingTimer()
            }

            WalletCore.Event.AccountConfigReceived -> {
                delegate.get()?.accountConfigChanged()
            }

            else -> {}
        }
    }

    override fun activityLoaderDataLoaded(isUpdateEvent: Boolean) {
        delegate.get()?.transactionsUpdated(isUpdateEvent = isUpdateEvent)
        dataUpdated()
        updateStatus()
    }

    override fun activityLoaderCacheNotFound() {
        delegate.get()?.cacheNotFound()
    }

    override fun activityLoaderLoadedAll() {
        delegate.get()?.loadedAll()
    }

    fun destroy() {
        stopUpdatingTimer()
        WalletCore.unregisterObserver(this)
    }
}
